### PR TITLE
A fiber can be a runnable

### PR DESCRIPTION
**Pull request by vasilmkd**
Related to #921.

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 1]
```diff
@@ -534,9 +556,10 @@ private[effect] final class IOFiber[A](
                     fiberB.registerListener(oc => cb(Right(Right((fiberA, oc)))))
 
                     val ec = currentCtx
-
-                    execute(ec)(() => fiberA.run(cur.ioa, ec))
-                    execute(ec)(() => fiberB.run(cur.iob, ec))
+                    fiberA.prepare(cur.ioa, ec)
+                    fiberB.prepare(cur.iob, ec)
+                    execute(ec)(fiberA)
+                    execute(ec)(fiberB)
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala:552]: LOL! I'm laughing so hard right now…

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 2]
```diff
@@ -499,12 +520,13 @@ private[effect] final class IOFiber[A](
             val childName = s"start-${childCount.getAndIncrement()}"
             val initMask2 = childMask
 
-            val fiber = new IOFiber(childName, scheduler, blockingEc, initMask2)
+            val fiber = new IOFiber[Any](childName, scheduler, blockingEc, initMask2)
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala:513]: Unnecessary change?

**Review by djspiewak — APPROVED** [review]
This is legit very clever. I'd like to hear if @RaasAhsan has any objections that I didn't notice, but it seems like a very smart way to cut down on the allocations in `start` and `race`

File: core/shared/src/main/scala/cats/effect/IOFiber.scala (hunk 2 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala:513]: I meant to ask this. There are two `IOFiber[Any]` instances created in `racePair`. This one here is actually an `IOFiber[Nothing]`. Does it need changing? Otherwise, I'll drop it.

Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala:513]: Oh, I didn't realize it was `IOFiber[Nothing]`. I'm cool with fixing that. The compiler sometimes freaks out around `Nothing`, so we may as well be explicit.

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 3]
```diff
@@ -187,6 +188,23 @@ private[effect] final class IOFiber[A](
...
+  private[this] var startIO: IO[Any] = _
+  private[this] var startEC: ExecutionContext = _
+
+  private def prepare(io: IO[Any], ec: ExecutionContext): Unit = {
+    startIO = io
+    startEC = ec
+  }
+
+  /**
+   * @note This method should not be used outside of the IO run loop under any circumstance.
```
Comment (RaasAhsan) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala]: I think it's fair to just declare the method straight up `private` then

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 4]
```diff
@@ -187,6 +188,23 @@ private[effect] final class IOFiber[A](
...
+   */
+  def run(): Unit = {
+    run(startIO, startEC)
```
Comment (RaasAhsan) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala]: maybe call this `start` now to avoid ambiguity

**Review by RaasAhsan — APPROVED** [review]
Very nice catch!

File: core/shared/src/main/scala/cats/effect/IOFiber.scala (hunk 3 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: That might not be possible due to the method coming from `Runnable`, but I'll try.

File: core/shared/src/main/scala/cats/effect/IOFiber.scala (hunk 4 above)
Comment (vasilmkd) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: Unfortunately it must be called `run`.  Should I put `override` to make it more clear?

Comment (RaasAhsan) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: ah `run(startIO, startEC)` this one needs to be called `run` as well?

Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: Let's rename that one. I'm not quite sure what to call it though.

Comment (vasilmkd) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: `exec`?

Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: I'm cool with `exec`
